docs: 4 class declarations (#5158)
dfinity/motoko: doc/md/fundamentals/3-declarations/4-class-declarations.md

---
sidebar_position: 4
---

# Class declarations

A class in Motoko serves as a blueprint for creating [objects](https://internetcomputer.org/docs/motoko/fundamentals/declarations/object-declaration) that encapsulate both [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior. It defines fields to hold data and methods to operate on that data. Unlike records and plain objects, classes support constructors, allowing developers to initialize each instance with unique values at creation time.

> Classes in Motoko are not the same as classes like in other object oritentated programming languages, but they serve the same purpose. Motoko also doesn’t have a `this` or `self` keyword because you can simply call other methods directly by name.

Reviewer: ```suggestion
Classes in Motoko are not the same as classes in other object oriented programming languages, but they serve the same purpose. Motoko also doesn’t have a `this` or `self` keyword because you can simply call other methods directly by name or name the entire object using an identifier of your choice.
```

 
## Defining a class

A class in Motoko defines both:

- A constructor for creating instances, which includes:
  - Fields to store data.
  - Methods to perform operations on that data.
- A type that describes the interface of the instance, including its available methods.

Motoko also supports generic classes, which allow abstraction over types. This enables the creation of reusable, type-safe components that can work with a variety of data types while maintaining strong compile-time guarantees.

```motoko no-repl
import O "mo:base/Order"; // Required to reference O.Order and its variants

  // A generic comparator class for ordering elements
  class Comparator<T>(compare : (T, T) -> O.Order) {
    public func lessThan(a : T, b : T) : Bool {
      compare(a, b) == #less;
    };

    public func equal(a : T, b : T) : Bool {
      compare(a, b) == #equal;
    };

    public func greaterThan(a : T, b : T) : Bool {
      compare(a, b) == #greater;
    };
  };

```

> This declaration creates:

Reviewer: ```suggestion
This declaration defines:
```

> - A constructor: `Comparator : ((K, K) -> Order) -> Comparator<K>`.

Reviewer: (missing type parameter)

- A generic type: `Comparator<K>` that provides comparison methods.

## Creating instances

Creating an instance is when a class constructor is used to produce a new object with its own state and behavior based on the class definition.

```motoko no-repl
class Counter(init : Nat) {
  var count : Nat = init;

  public func inc() : Nat {
    count += 1;
    count
  };
};

// Creating an instance of the class
let myCounter = Counter(0);

// Using the instance
myCounter.inc(); // returns 1
```

## Actor classes

Actor classes enable you to create networks of actors programmatically. These classes must be defined in separate source files.

```motoko no-repl title="Counter.mo"
actor class Counter() {
  var count : Int = 0;

  public func increment() : async Int {
    count += 1;
    return count;
  };

  public func get() : async Int {
    return count;
  };
}
```

This defines a `Counter` actor class with:

- A mutable variable `count`.

- A method `increment` that increases `count` and returns the new value.

> - A method get that returns the current count.

Reviewer: ```suggestion
- A method `get` that returns the current count.
```

Then, in another file you can create an instance of this actor class:

```motoko no-repl title="CallCounter.mo"
import Counter "./Counter";

actor {

>   let counter = await (with cycles = 1_000_000_000_000) Counter.Counter();

Reviewer: This shouldn't typecheck since you can't await or sent messages in the body of an actor.

See the suggested rewrite below.

  let newCount = await counter.increment();

  Debug.print("Count is now: " # Int.toText(newCount));

  let currentCount = await counter.get();
  Debug.print("Current count: " # Int.toText(currentCount));
}
```

> 

Reviewer: ```suggestion

Unlike object classes, actor class constructors are asynchronous: the constructor returns a future that contains the instance. You obtain the instance itself using `await`.

```

myCounter.inc(); // returns 1
```

The class declaration syntax also lets you:
* Check that the class conforms to some interface, expressed as an object type.
* Define a self-identifier to refer to the entire object.

```motoko no-repl
type Reset = { reset : () -> () };

class Counter(init : Nat) : Reset = this {
  var count : Nat = init;

  public func inc() : Nat {
    count += 1;
    count
  };
  
  public func getThis() : Counter { this };
    
  public func reset() { count := 0 };
};
```

> This extended `Counter` class checks that the implementation matches interface `Reset` and names this parameter `self`.

Reviewer: ```suggestion
This extended `Counter` class checks that the implementation matches interface `Reset` and names this parameter `self`.

```
